rpc/src/connection.rs

```diff
@@ -0,0 +1,529 @@
+
+impl Default for OperationCallback {
+    fn default() -> Self {
+        Self {
+            data: std::ptr::null(),
+            meta_data: std::ptr::null(),
+            data_length: 0,
+            meta_data_length: 0,
+            state: CallbackState::Empty,
+            request_status: 0,
+            flags: 0,
+            channel: mpsc::channel(),
+        }
+    }
+}
+
+impl OperationCallback {
+    pub fn new() -> Self {
+        Self {
+            data: std::ptr::null(),
+            meta_data: std::ptr::null(),
+            data_length: 0,
+            meta_data_length: 0,
+            state: CallbackState::Empty,
+            request_status: 0,
+            flags: 0,
+            channel: mpsc::channel(),
+        }
+    }
+}
+
+pub struct CircularQueue {
+    callbacks: Vec<*const OperationCallback>,
+    start_index: Arc<Mutex<u32>>, // maybe we can use a lock-free queue
+    end_index: Arc<Mutex<u32>>,
+}
+
+impl Default for CircularQueue {
+    fn default() -> Self {
+        Self {
+            callbacks: vec![std::ptr::null(); REQUEST_QUEUE_LENGTH],
+            start_index: Arc::new(Mutex::new(0)),
+            end_index: Arc::new(Mutex::new(1)),
+        }
+    }
+}
+
+impl CircularQueue {
+    pub fn new() -> Self {
+        Self {
+            callbacks: vec![std::ptr::null_mut(); REQUEST_QUEUE_LENGTH],
+            start_index: Arc::new(Mutex::new(0)),
+            end_index: Arc::new(Mutex::new(1)),
+        }
+    }
+
+    pub fn init(&mut self) {
+        for i in 0..self.callbacks.len() {
+            // allocate memory for the callback on the heap to avoid thread context switch.
+            // ensure that the callback is not moved by the compiler.
+            self.callbacks[i] = Box::into_raw(Box::new(OperationCallback::new()));
+        }
+        *self.end_index.lock().unwrap() = 0;
+    }
+
+    pub fn register_callback(
+        &self,
+        rsp_data: &mut [u8],
+        rsp_meta_data: &mut [u8],
+    ) -> Result<u32, Box<dyn std::error::Error>> {
+        let mut end_index = self.end_index.lock().unwrap();
+        let id = *end_index;
+        *end_index = (*end_index + 1) % REQUEST_QUEUE_LENGTH as u32;
+        unsafe {
+            let callback = self.callbacks[id as usize];
+            (*(callback as *mut OperationCallback)).state = CallbackState::WaitingForResponse;
+            (*(callback as *mut OperationCallback)).data = rsp_data.as_ptr();
+            (*(callback as *mut OperationCallback)).meta_data = rsp_meta_data.as_ptr();
+            (*(callback as *mut OperationCallback)).data_length = rsp_data.len();
+            (*(callback as *mut OperationCallback)).meta_data_length = rsp_meta_data.len();
+        }
+        Ok(id)
+    }
+
+    pub fn clean_up(&self) -> Result<(), Box<dyn std::error::Error>> {
+        let mut start_index = self.start_index.lock().unwrap();
+        let end_index = self.end_index.lock().unwrap();
+        let end_flag = *end_index;
+        drop(end_index);
```

Comment:
why drop?

Reply:
Other thread may use self.end_index, so we need to release the lock.